activesupport/lib/active_support/cache/coder.rb

```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+require_relative "entry"
+
+module ActiveSupport
+  module Cache
+    class Coder # :nodoc:
+      def initialize(serializer, compressor, legacy_serializer: false)
+        @serializer = serializer
+        @compressor = compressor
+        @legacy_serializer = legacy_serializer
+      end
+
+      def dump(entry)
+        return @serializer.dump(entry) if @legacy_serializer
+
+        dump_compressed(entry, Float::INFINITY)
+      end
+
+      def dump_compressed(entry, threshold)
+        return @serializer.dump_compressed(entry, threshold) if @legacy_serializer
+
+        # If value is a string with a supported encoding, use it as the payload
+        # instead of passing it through the serializer.
+        if type = type_for_string(entry.value)
+          payload = binary_compatible_string(entry.value)
+        else
+          type = OBJECT_DUMP_TYPE
+          payload = serialize(entry.value)
+        end
+
+        # Compress payload if it meets the size threshold. (Note that `type < 0`
+        # signifies a compressed payload with `type.abs` as its actual type.)
+        if compressed = try_compress(payload, threshold)
+          payload = compressed
+          type = -type
+        end
+
+        expires_at = entry.expires_at || -1.0
+
+        version = dump_version(entry.version)
+        version_length = version&.bytesize || -1
+
+        packed = [SIGNATURE, type, expires_at, version_length].pack(PACKED_TEMPLATE)
+        packed << version if version
+        packed << payload
+      end
+
+      def load(dumped)
+        return @serializer.load(dumped) if !signature?(dumped)
+
+        type = dumped.unpack1(PACKED_TYPE_TEMPLATE)
+        expires_at = dumped.unpack1(PACKED_EXPIRES_AT_TEMPLATE)
+        version_length = dumped.unpack1(PACKED_VERSION_LENGTH_TEMPLATE)
+
+        expires_at = nil if expires_at < 0
+        version = load_version(dumped.byteslice(PACKED_VERSION_INDEX, version_length))
```

Comment:
```suggestion
        version = load_version(dumped.byteslice(PACKED_VERSION_INDEX, version_length)) if version_length >= 0
```

I think I'd prefer if we explicitly handled negative length here. Lots of cache payloads don't have a version at all so might as well skip `load_version` and it also allow to remove the nil check in `load_version`.